contracts/BC_fusion/StakeHub.sol

```diff
@@ -0,0 +1,672 @@
+
+        emit Redelegated(srcValidator, dstValidator, delegator, _bnbAmount);
+
+        bytes memory srcValVoteAddr = srcValInfo.voteAddress;
+        bytes memory dstValVoteAddr = dstValInfo.voteAddress;
+        _updateEligibleValidators(UpdateDirection.Down, srcValidator, srcValVoteAddr);
+        _updateEligibleValidators(UpdateDirection.Up, dstValidator, dstValVoteAddr);
+    }
+
+    function claimSecurityDeposit(uint256 _sharesAmount) external onlyInitialized validatorExist(msg.sender) {
+        Validator memory valInfo = validators[msg.sender];
+        address pool = valInfo.poolModule;
+        require(IStakePool(pool).balanceOf(address(this)) >= _sharesAmount, "NOT_ENOUGH_SHARES");
+
+        emit SecurityDepositClaimed(msg.sender, _sharesAmount);
+
+        IStakePool(pool).transfer(msg.sender, _sharesAmount);
+        if (IStakePool(pool).getSecurityDepositBNB() < minSelfDelegationBNB) {
+            valInfo.jailed = true;
+            emit ValidatorJailed(msg.sender);
+        }
+    }
+
+    function claim(address validator, uint256 requestNumber) external onlyInitialized validatorExist(validator) {
+        uint256 _bnbAmount = IStakePool(validators[validator].poolModule).claim(msg.sender, requestNumber);
+
+        emit Claimed(validator, msg.sender, _bnbAmount);
+    }
+
+    /*----------------- system functions -----------------*/
+    function distributeReward(address validator) external payable onlyInitialized onlyValidatorContract validatorNotJailed(validator) {
+        IStakePool(validators[validator].poolModule).distributeReward{value: msg.value}();
+    }
+
+    function getEligibleValidators() external view returns (IBSCValidatorSet.Validator[] memory, bytes[] memory) {
+        return (eligibleValidators, eligibleValidatorVoteAddrs);
+    }
+
+    function downtimeSlash(address validator, uint256 height) external onlyInitialized onlySlash {
+        Validator storage valInfo = validators[validator];
+
+        // check slash record
+        bytes32 slashKey = _getSlashKey(validator, height, SlashType.DownTime);
+        SlashRecord storage record = slashRecords[slashKey];
+        require(record.slashHeight == 0, "SLASHED");
+
+        // slash
+        uint256 slashAmount = IStakePool(valInfo.poolModule).slash(downtimeSlashAmount);
+        valInfo.jailed = true;
+        emit ValidatorJailed(validator);
+
+        // record
+        record.slashAmount = slashAmount;
+        record.slashHeight = height;
+        record.jailUntil = block.timestamp + downtimeJailTime;
+        record.slashType = SlashType.DownTime;
+
+        if (valInfo.jailUntil < record.jailUntil) {
+            valInfo.jailUntil = record.jailUntil;
+        }
+
+        emit ValidatorSlashed(validator, slashAmount, height, record.jailUntil, SlashType.DownTime);
+    }
+
+    function maliciousVoteSlash(bytes calldata voteAddr, uint256 height) external onlyInitialized onlySlash {
```

Comment:
Do we need to do `_updateEligibleValidators` when slash happened?

Reply:
Sure. I forgot this.